src/main/java/pl/betoncraft/betonquest/utils/location/AbstractData.java

```diff
@@ -0,0 +1,124 @@
+     */
+    public AbstractData(final String packName, final String data) throws InstructionParseException {
+        final Matcher variableMatcher = PATTERN_VARIABLE.matcher(data);
+        if (variableMatcher.find()) {
+            objectVariables = new ArrayList<>();
+            int index = 1;
+            final StringBuffer stringBuffer = new StringBuffer(data.length());
+            do {
+                final String variable = variableMatcher.group(0);
+                final Variable var = BetonQuest.createVariable(Config.getPackages().get(packName), variable);
+                objectVariables.add(var);
+                final String replacement = "%" + index++ + "$s";
+                variableMatcher.appendReplacement(stringBuffer, Matcher.quoteReplacement(replacement));
+            } while (variableMatcher.find());
+            variableMatcher.appendTail(stringBuffer);
+            objectFormatted = stringBuffer.toString();
+            object = null;
+        } else {
+            try {
+                object = parse(data);
+            } catch (final InstructionParseException e) {
+                throw new InstructionParseException("Error while parsing vector. " + e.getMessage(), e);
+            }
+            objectVariables = null;
+            objectFormatted = null;
+        }
+    }
+
+    /**
+     * Parses a string with resolved {@link Variable}s into a {@link T}.
+     *
+     * @param objectString The string that represents the vector
+     * @return The parsed {@link T}
+     * @throws InstructionParseException Is thrown when the objectString is not in the right format or if
+     *                                   the values couldn't be parsed.
+     */
+    public abstract T parse(final String objectString) throws InstructionParseException;
+
+    /**
+     * Get the {@link T}
+     *
+     * @param playerID ID of the player - needed for {@link Variable}s resolution
+     * @return The {@link T} represented by this object
+     * @throws QuestRuntimeException Is thrown when the objectString is not in the right format or if
+     *                               the values couldn't be parsed.
+     */
+    public T get(final String playerID) throws QuestRuntimeException {
+        return object == null ? parseVariableObject(playerID) : object;
+    }
+
+    private T parseVariableObject(final String playerID) throws QuestRuntimeException {
+        if (playerID == null) {
+            throw new QuestRuntimeException("Variable cannot be accessed without the player."
+                    + " Consider changing it to absolute coordinates");
+        }
+        final String[] variables = new String[this.objectVariables.size()];
+        for (int i = 0; i < this.objectVariables.size(); i++) {
+            final Variable var = this.objectVariables.get(i);
+            variables[i] = var.getValue(playerID);
+        }
+        final String result = String.format(objectFormatted, (Object[]) variables);
```

Comment:
Is casting here strictly required? Maybe even declare `variables` as `Object[]` to begin with?

Reply:
yes